src/backend/distributed/utils/node_metadata.c

```diff
@@ -0,0 +1,722 @@
+
+	/* check if the node already exists in the cluster */
+	workerNode = FindWorkerNode(nodeName, nodePort);
+	if (workerNode != NULL)
+	{
+		/* fill return data and return */
+		returnData = GenerateNodeTuple(workerNode);
+
+		/* close the heap */
+		heap_close(pgDistNode, AccessExclusiveLock);
+
+		PG_RETURN_DATUM(returnData);
+	}
+
+	/* user lets Citus to decide on the group that the newly added node should be in */
+	if (groupId == 0)
+	{
+		groupId = NextGroupId();
+	}
+	else
+	{
+		uint maxGroupId = GetMaxGroupId();
+
+		if (groupId > maxGroupId)
+		{
+			ereport(ERROR, (errmsg("you cannot add a node to a non-existing group")));
+		}
+	}
+
+	/* generate the new node id from the sequence */
+	nextNodeId = master_get_next_nodeid(NULL);
+	nextNodeIdInt = DatumGetUInt32(nextNodeId);
+
+	InsertNodeRow(nextNodeIdInt, nodeName, nodePort, groupId, nodeRack);
+
+	heap_close(pgDistNode, AccessExclusiveLock);
+
+	/* fetch the worker node, and generate the output */
+	workerNode = FindWorkerNode(nodeName, nodePort);
+	returnData = GenerateNodeTuple(workerNode);
+
+	return returnData;
+}
+
+
+/*
+ * GenerateNodeTuple gets a worker node and return a heap tuple of
+ * given worker node.
+ */
+static Datum
+GenerateNodeTuple(WorkerNode *workerNode)
+{
+	Relation pgDistNode = NULL;
+	TupleDesc tupleDescriptor = NULL;
+	HeapTuple heapTuple = NULL;
+	Datum nodeDatum = 0;
+	Datum values[Natts_pg_dist_node];
+	bool isNulls[Natts_pg_dist_node];
+
+	/* form new shard tuple */
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+
+	values[Anum_pg_dist_node_nodeid - 1] = UInt32GetDatum(workerNode->nodeId);
+	values[Anum_pg_dist_node_groupid - 1] = UInt32GetDatum(workerNode->groupId);
+	values[Anum_pg_dist_node_nodename - 1] = CStringGetTextDatum(workerNode->workerName);
+	values[Anum_pg_dist_node_nodeport - 1] = UInt32GetDatum(workerNode->workerPort);
+	values[Anum_pg_dist_node_noderack - 1] = CStringGetTextDatum(workerNode->workerRack);
+
+	/* open shard relation and insert new tuple */
+	pgDistNode = heap_open(DistNodeRelationId(), AccessShareLock);
+
+	/* generate the tuple */
+	tupleDescriptor = RelationGetDescr(pgDistNode);
+	heapTuple = heap_form_tuple(tupleDescriptor, values, isNulls);
+
+	nodeDatum = HeapTupleGetDatum(heapTuple);
+
+	/* close the relation */
+	heap_close(pgDistNode, AccessShareLock);
+
+	return nodeDatum;
+}
+
+
+/*
+ * NextGroupId returns the next group that that can be assigned to a node. If the
+ * group is full (i.e., it has equal or more elements than GroupSize), a new group id
+ * is generated and returned. Else, the current maximum group id is returned.
+ */
+static uint32
+NextGroupId()
+{
+	uint32 nextGroupId = 0;
+	uint32 maxGroupIdInt = GetMaxGroupId();
+	uint64 nodeCountInMaxGroupId = GetNodeCountInGroup(maxGroupIdInt);
+
+	if (nodeCountInMaxGroupId == 0 || nodeCountInMaxGroupId >= GroupSize)
+	{
+		Datum nextGroupIdDatum = master_get_next_groupid(NULL);
```

Comment:
Calling a UDF should use DirectFunctionCall1 (see below). Alternatively, we could just turn them into regular functions.